azkaban-common/src/main/java/azkaban/executor/mail/MailCreatorRegistry.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2012 LinkedIn Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package azkaban.executor.mail;
+
+import org.apache.log4j.Logger;
+
+import java.util.HashMap;
+
+public class MailCreatorRegistry {
+
+  private static final Logger logger = Logger.getLogger(MailCreatorRegistry.class);
+  private static final HashMap<String, MailCreator> registeredCreators = new HashMap<>();
+  private static final MailCreator defaultCreator;
+  private static MailCreator recommendedCreator;
+
+  static {
+    recommendedCreator = defaultCreator = new DefaultMailCreator();
+    registerCreator(defaultCreator, true);
+  }
+
+  public static void registerCreator(final MailCreator creator, final boolean recommend) {
+    logger.info("registering mail creator " + creator.getName() + ", recommended = " + recommend);
+    registeredCreators.put(creator.getName(), creator);
+    if (recommend) {
```

Comment:
I don't fully understand why you need to introduce the recommendedCreator. 
It seems everywhere this registerCreator method is called, the recommend flag passed is always true. In what scenario would you set the recommend flag to false? 
If we could decide the creator during server initialization like you proposed, would it be possible to simplify the logic?

Reply:
The discussion I can see here is about the clarity of intent/design. The `AzkabanServer` class is currently the only one that registers a mail creator. It is hard for me to foresee whether there can be another such component and what would be the desired logic to satisfy their potential claims. The flag is one way to give registering side some control. A more straightforward approach would be that the last registered mail creator wins/is recommended for the rest of the system to use.
I am updating the PR to show this simpler approach would look.